### PR TITLE
adding BenchmarkDotNet support for CoreFX purpose

### PR DESCRIPTION
**Pull request by adamsitnik**
We are currently transitioning from xunit-performance to BenchmarkDotNet.

All of the benchmarks live in the performance repo, however when CoreFX developers want to run benchmarks against new API that is not exposed by any contracts yet they need to test it locally.

This is why I am adding support for BDN to CoreFX. Before I do that, I need these changes to merged and propagated ofc ;)

Some reasoning behind the changes:

1. All changes are inside a `'$(IsBenchmarkDotNetProject)' == 'true'` condition to make sure I don't break anybody
2. I did not remove xunit-performance stuff, just added BDN

How it's going to work in the CoreFX repo for running benchmarks against new API: exactly the same way it was with xunit-performance:

Default run:

> dotnet msbuild /t:BuildAndTest /p:ConfigurationGroup=Release /p:Performance=true

Diagnostics run with same Hardware Counters as xunit so far:

> dotnet msbuild /t:BuildAndTest /p:ConfigurationGroup=Release /p:Performance=true /p:PerformanceType=Diagnostic

New things: possibility to filter benchmarks using `/p:PerfRunnerFilter`:

> dotnet msbuild /t:BuildAndTest /p:ConfigurationGroup=Release /p:Performance=true /p:PerfRunnerFilter=*MyFilter*

Pass some extra BDN parameters using `/p:PerfRunnerArgs`:

dotnet msbuild /t:BuildAndTest /p:ConfigurationGroup=Release /p:Performance=true /p:PerfRunnerArgs=--join

I tested all of the combindations, including `/p:CoreCLROverridePath`. It works.

I am going to create the CoreFX PR as soon as @ViktorHofer gives me his MSBuild blessing for these changes ;)

/cc @danmosemsft

### CODE REVIEW

**Comment by ViktorHofer** [comment]
The changes overall look good but I would need the corresponding corefx changes for context to review it carefully.

**Comment by adamsitnik** [comment]
@ViktorHofer https://github.com/dotnet/corefx/pull/32843

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/PerfTesting.targets [hunk 1]
```diff
@@ -73,6 +73,26 @@
...
+    <ProfilerName Condition="'$(TargetOS)'=='Windows_NT'">ETW</ProfilerName>
+    <!-- by default there are no args, this can be used by the users to pass BenchmarkDotNet arguments here -->
+    <PerfRunnerArgs></PerfRunnerArgs>
+    <!-- when we run a Diagnostic benchmark run we specify the profiler and the hardware counters we want to get -->
+    <PerfRunnerArgs Condition="'$(PerformanceType)'=='Diagnostic'">--profiler $(ProfilerName) --counters $(CollectFlags)</PerfRunnerArgs>
+    <PerfRunnerAllArgs>--filter $(PerfRunnerFilter) $(PerfRunnerArgs) || $(CliExitErrorCommand)</PerfRunnerAllArgs>
+    <PerfRunnerCommand Condition="'$(BuildingNETFxVertical)' != 'true'">$(PerfTestCommandDotnetExecutable) $(PerfRunnerExecutable) $(PerfRunnerAllArgs)</PerfRunnerCommand>
+    <PerfRunnerCommand Condition="'$(BuildingNETFxVertical)' == 'true'">$(PerfRunnerExecutable) $(PerfRunnerAllArgs)</PerfRunnerCommand>
+    <MeasurementPyCommand>$(PythonCommand) "$(BenchviewDir)/measurement.py" xunit "$(RunId)-$(AssemblyName).xml" --better desc --append -o "$(ProjectDir)measurement.json" || $(CliExitErrorCommand)</MeasurementPyCommand>
```
Comment (jorive) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/PerfTesting.targets]: Is this line needed?

**Comment by adamsitnik** [comment]
@ViktorHofer I am going to merge this one to be able to clean up, test and finish #32843

I have verified that I don't need any other changes in the .targets files

**Comment by ViktorHofer** [comment]
Works for me, thanks for the work. We are right now in the middle of migrating the whole testing infrastructure to arcade. I'm close to submitting a PR and will add your changes into it.

**Comment by adamsitnik** [comment]
@ViktorHofer thanks for letting me know! it means that this particular change is not going to be merged OOB by maestro bot but as a part of a bigger PR?

**Comment by ViktorHofer** [comment]
Right it requires two PRs: one in arcade and one in corefx to consume the arcade package. I will cc you to review the corresponding perf changes.

**Pull request merged**